servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SingleProcessor.java

```diff
@@ -58,9 +58,16 @@ protected void handleSubscribe(final Subscriber<? super T> subscriber) {
 
         if (subscribers.offer(subscriber)) {
             Object terminalSignal = this.terminalSignal;
-            if (terminalSignal != TERMINAL_NULL) {
-                // To ensure subscribers are notified in order we go through the queue to notify subscribers.
-                notifyListeners(terminalSignal);
+            if (terminalSignal != TERMINAL_NULL && subscribers.remove(subscriber)) {
+                if (terminalSignal instanceof TerminalNotification) {
+                    final Throwable error = ((TerminalNotification) terminalSignal).cause();
+                    assert error != null : "Cause can't be null from TerminalNotification.error(..)";
+                    subscriber.onError(error);
```

Comment:
should we use `safeOnError` and `safeOnSuccess` here? Otherwise the Single may catch the throwable and propagate another duplicate terminal event.

Reply:
also while we are here perhaps consider using `try { subscriber.onSubscribe(..); } catch(e) { handleExceptionFromOnSubscribe(..) } ` for more graceful error handling.